src/Processors/QueryPlan/ReadFromRemote.cpp

```diff
@@ -151,6 +168,249 @@ void ReadFromRemote::enforceAggregationInOrder()
+        {
+            node_to_ast[node] = nullptr;
+            continue;
+        }
+
+        /// Support for IN. The stored AST from the Set is taken.
+        if (WhichDataType(node->result_type).isSet())
+        {
+            auto maybe_set = node->column;
+            if (const auto * col_const = typeid_cast<const ColumnConst *>(maybe_set.get()))
+                maybe_set = col_const->getDataColumnPtr();
+
+            if (const auto * col_set = typeid_cast<const ColumnSet *>(maybe_set.get()))
+                node_to_ast[node] = col_set->getData()->getSourceAST();
+
+            continue;
+        }
+
+        if (node->column && isColumnConst(*node->column))
+        {
+            auto literal = std::make_shared<ASTLiteral>((*node->column)[0]);
+            node_to_ast[node] = std::move(literal);
+            stack.pop();
+            continue;
+        }
+
+        if (frame.next_child < node->children.size())
+        {
+            stack.push({node->children[frame.next_child]});
+            ++frame.next_child;
+            continue;
+        }
+
+        stack.pop();
+        auto & res = node_to_ast[node];
+
+        if (node->type == ActionsDAG::ActionType::INPUT)
+        {
+            /// The column name can be taken from the projection name, or from the projection expression.
+            /// It depends on the predicate and query stage.
+
+            if (projection_names.contains(node->result_name))
+            {
+                /// The input name matches the projection name. Example:
+                /// SELECT x FROM (SELECT number + 1 AS x FROM remote('127.0.0.2', numbers(3))) WHERE x = 1
+                /// In this case, ReadFromRemote has header `x UInt64` and filter DAG has input column with name `x`.
+                /// Here, filter is applied to the whole query, and checking for projection name is reasonable.
+                res = std::make_shared<ASTIdentifier>(node->result_name);
+            }
+            else
+            {
+                /// The input name matches the execution name of the projection. Example:
+                /// SELECT x, y FROM (
+                ///     SELECT number + 1 AS x, sum(number) AS y FROM remote('127.0.0.{1,2}', numbers(3)) GROUP BY x
+                /// ) WHERE x = 1
+                /// In this case, ReadFromRemote has `plus(__table1.number, 1_UInt8) UInt64` in header,
+                /// and filter DAG has input column with name `plus(__table1.number, 1_UInt8)`.
+                /// Here, filter is pushed down before the aggregation, and projection name can't be used.
+                /// However, we can match the input with the execution name of projection query tree.
+                /// Note: this may not cover all the cases.
+                auto it = execution_name_to_projection_query_tree.find(node->result_name);
+                if (it != execution_name_to_projection_query_tree.end())
```

Comment:
Is it possible to not find it?

Reply:
I hope not, but I could not prove it.
Skipping the optimization feels safer.